src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionXmlFormatReader.cs

```diff
@@ -465,9 +576,91 @@ private bool ReflectionTryReadPrimitiveArray(Type type, Type itemType, int array
+            {
+                object resultCollection = _arg1Context.GetCollectionMember();
+                if (resultCollection == null)
+                {
+                    XmlObjectSerializerReadContext.ThrowNullValueReturnedForGetOnlyCollectionException(collectionContract.UnderlyingType);
+                }
+
+                Type collectionType = collectionContract.UnderlyingType;
+                bool isArray = (collectionContract.Kind == CollectionKind.Array);
+                int arraySize = 0;
+                if (isArray)
+                {
+                    arraySize = ((Array)resultCollection).Length;
+                }
+
+                string itemNameStr = collectionContract.ItemName;
+                string itemNsStr = collectionContract.StableName.Namespace;
+
+                for (int index = 0; index < int.MaxValue; index++)
+                {
+                    if (_arg0XmlReader.IsStartElement(_arg2CollectionItemName, _arg3CollectionItemNamespace))
+                    {
+                        _arg1Context.IncrementItemCount(1);
+                        object collectionItem = ReflectionReadCollectionItem(collectionContract, collectionContract.ItemType, itemNameStr, itemNsStr);
+                        if (isArray)
+                        {
+                            if (index == arraySize)
+                            {
+                                XmlObjectSerializerReadContext.ThrowArrayExceededSizeException(arraySize, collectionType);
+                            }
+                            
+                            ((Array)resultCollection).SetValue(collectionItem, index);
+                        }
+                        else if (collectionContract.Kind == CollectionKind.GenericDictionary || collectionContract.Kind == CollectionKind.Dictionary)
+                        {
+                            var getKeyMethod = collectionContract.ItemType.GetMethod("get_Key");
+                            var getValueMethod = collectionContract.ItemType.GetMethod("get_Value");
+                            object key = getKeyMethod.Invoke(collectionItem, emptyObjectArray);
+                            object value = getValueMethod.Invoke(collectionItem, emptyObjectArray);
+                            IDictionary dict = (IDictionary)resultCollection;
+                            dict.Add(key, value);
+                        }
+                        else
+                        {
+                            Type genericCollectionType = typeof(ICollection<>).MakeGenericType(collectionContract.ItemType);
+                            Type typeIList = typeof(IList);
+                            if (genericCollectionType.IsAssignableFrom(collectionType))
+                            {
+                                MethodInfo addMethod = typeof(ICollection<>).MakeGenericType(collectionContract.ItemType).GetMethod("Add");
```

Comment:
I haven't done the analysis in how fast this is.  Matt, do you know whether it would be faster to keep a cache of created generic types of ICollection<T> and not do it everytime?